source/ada/lsp-ada_handlers.adb

```diff
@@ -2022,6 +2026,158 @@ package body LSP.Ada_Handlers is
+                      Context.Analysis_Units;
+               begin
+                  Last := Index + Context_Units'Length - 1;
+                  Units (Index ..  Last) := Context_Units;
+                  Index := Last + 1;
+               end;
+            end loop;
+
+            return Unit.P_Filter_Is_Imported_By
+              (Units      => Units,
+               Transitive => Params.showImplicit);
+         end;
+      end Get_Importing_Units;
+
+      procedure Append_Units
+        (Units : Libadalang.Analysis.Analysis_Unit_Array);
+      procedure Append_Units
+        (Units : Libadalang.Analysis.Compilation_Unit_Array);
+
+      ------------------
+      -- Append_Units --
+      ------------------
+
+      procedure Append_Units
+        (Units : Libadalang.Analysis.Analysis_Unit_Array) is
+      begin
+         for Unit of Units loop
+            Response.result.Append
+              (LSP.Messages.ALS_Unit_Description'
+                 (uri        => From_File
+                      (GNATCOLL.VFS.Create (+Unit.Get_Filename)),
+                  projectUri => From_File (Self.Root)));
+         end loop;
+      end Append_Units;
+
+      procedure Append_Units
+        (Units : Libadalang.Analysis.Compilation_Unit_Array) is
+      begin
+         for Unit of Units loop
+            Response.result.Append
+              (LSP.Messages.ALS_Unit_Description'
+                 (uri        => From_File
+                      (GNATCOLL.VFS.Create (+Unit.Unit.Get_Filename)),
+                  projectUri => From_File (Self.Root)));
+         end loop;
+      end Append_Units;
+
+   begin
+
+      case Root.Kind is
+         when Libadalang.Common.Ada_Compilation_Unit_Range =>
+            declare
+               Comp_Unit    : constant Compilation_Unit :=
+                                As_Compilation_Unit (Root);
+
+            begin
+               if Params.kind = LSP.Messages.Show_Imported then
+                  declare
+                     Units : constant Compilation_Unit_Array :=
+                       (if Params.showImplicit then
+                           Comp_Unit.P_Unit_Dependencies
+                        else
+                           Comp_Unit.P_Withed_Units);
+                  begin
+                     Append_Units (Units);
+                  end;
+               else
+                  declare
+                     Units : constant Libadalang.Analysis.Analysis_Unit_Array
+                       := Get_Importing_Units (Comp_Unit);
+                  begin
+                     Append_Units (Units);
+                  end;
+               end if;
+            end;
+
+         when Libadalang.Common.Ada_Compilation_Unit_List_Range =>
+            declare
+               Comp_Unit_List : constant Compilation_Unit_List :=
+                                As_Compilation_Unit_List (Root);
+            begin
+               for Comp_Unit of Comp_Unit_List loop
```

Comment:
Looks like copy/paste of previous case.

Reply:
No it's not because the type is different (Compilation_Unit vs Compilation_Unit_List). I ddid not find any proper solution to factorize the code unfortunately.